### ISSUE TITLE
metadata object has unspecified API which leaks EmailMessage implementation

### ISSUE THREAD

**Issue by jaraco** [python/importlib_metadata#126]
In GitLab by [@eschwartz](https://gitlab.com/eschwartz) on Jun 16, 2020, 15:01

Noted on IRC,

> Technically, the returned distribution metadata object is an email.message.EmailMessage instance, but this is an implementation detail, and not part of the stable API. You should only use dictionary-like methods and syntax to access the metadata contents.

According to this definition, something like `.get_all()` should not be relied upon, because it is not a dictionary method. But `.get_all()` serves a valuable role, see for example the implementation of `_read_dist_info_reqs()`, which retrieves multiple definitions of `Requires-Dist` to define the `Distribution.requires` property.

With only access to `metadata['Requires-Dist']` it's impossible to get reliable information for metadata that can have multiple values. With `metadata.items()` it's possible to get this information, but then you'd need to filter it yourself, which is clumsy and inefficient.

    <jaraco> What the project needs is: A test capturing the need for multi-valued keys, a protocol describing the interface that 'metadata' guarantees to provide, and then an update to the docs to point to the protocol as the proper interface such that EmailMessage need not be mentioned.

tl;dr it seems like a good idea to document `.get_all()` and anything else generally useful as part of the public `Distribution.metadata`, API, without mentioning `EmailMessage` at all.

**Comment by jaraco** [comment]
In GitLab by [@FFY00](https://gitlab.com/FFY00) on Jun 16, 2020, 15:39

I believe `EmailMessage` should still be mentioned as otherwise people will assume the other methods are just missing from the documentation.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Sep 22, 2020, 20:13

In 5e7d0d8, I've attempted to add an interface definition on the return type for metadata. Unfortunately, it causes this error from mypy:

```
importlib_metadata/__init__.py:287: error: Cannot call function of unknown type
```

It seems the `email_message_from_string` function is an unknown type. The [definition for that function](https://gitlab.com/python-devs/importlib_metadata/-/blob/46efbf9df587e0a799fa6166dbc056486e02d106/importlib_metadata/_compat.py#L107-118) is selective on Python version, so I'm not even sure mypy can discern the return type for the different forms that function might take. I tried placing a `cast(PackageMetadata, email_message_from_string(text))`, but that resulted in the same error.

Does anyone with more familiarity with mypy know how to declare the types in this situation?

**Comment by FFY00** [comment]
That does not work because `collections.abc` defines base classes, this means our object should be subclasses from the type. I am not really sure why casting doesn't work.

The solution is to use `typing.Protocol`, which lets us define the public API of the object but objects do not actually need to be subclassed from it.

**Comment by FFY00** [comment]
With #263 merged, this can now be closed.
